language/move-prover/tests/sources/mut_ref_accross_modules.move

```diff
@@ -3,53 +3,107 @@ module TestMutRefs {
 
     struct T { value: u64 }
 
+    // Resource to track the sum of values in T
     resource struct TSum {
         sum: u64
     }
 
     spec struct T {
+        // global specification variable tracking sum of values.
         global spec_sum: u64;
 
-        invariant update value > old(value);
+        // Data invariant.
+        invariant value > 0;
+
+        // Pack/unpack invariants updating spec var
         invariant pack spec_sum = spec_sum + value;
         invariant unpack spec_sum = spec_sum - value;
     }
 
     spec module {
+        // Module invariant that the tracked resource sum equals the spec sum
         invariant global<TSum>(0x0).sum == spec_sum;
     }
 
+    // This succeeds because module invariant can be assumed on entry (public function) and is maintained on exit.
     public fun new(x: u64): T acquires TSum {
         let r = borrow_global_mut<TSum>(0x0);
         r.sum = r.sum + x;
         T{value: x}
     }
+    spec fun new {
+        requires x > 0;
+    }
 
+    // This succeeds because module invariant can be assumed on entry (public function) and is maintained on exit.
     public fun delete(x: T) acquires TSum {
         let r = borrow_global_mut<TSum>(0x0);
         let T{value: v} = x;
         r.sum = r.sum - v;
     }
 
+    // This succeeds because module invariant can be assumed on entry (public function) and is maintained on exit.
     public fun increment(x: &mut T) acquires TSum {
         x.value = x.value + 1;
         let r = borrow_global_mut<TSum>(0x0);
         r.sum = r.sum + 1;
     }
 
-    // This should fail because the update invariant requires strict monotonic increase of the value.
-    // TODO: we see this error only when we comment out the next function. Need to determine where this
-    //   comes from, as we usually see Boogie reporting all errors, and not aborting after the first one.
+    // This should fail because we do not update the TSum resource.
+    public fun increment_invalid(x: &mut T) {
+        x.value = x.value + 1;
+    }
+
+    // This should fail because we violate the data invariant (value always > 0). The data invariant is enforced
+    // on exit even though it works on a mut ref since this is a public method.
     public fun decrement_invalid(x: &mut T) acquires TSum {
         x.value = x.value - 1;
         let r = borrow_global_mut<TSum>(0x0);
         r.sum = r.sum - 1;
     }
 
-    // This should fail because we do not update the TSum resource.
-    public fun increment_invalid(x: &mut T) {
-        x.value = x.value + 1;
+    // This one is exactly the same as the previous but succeeds since it is private. The data invariant is not
+    // enforced because the mut ref is manipulated privately.
```

Comment:
At the same time, the data invariant should not be assumed upon entry either. So if I wrote spec {assert x.value > 0} at the very beginning of this function, I should get an error.

Reply:
Good point. This is exactly #3076. I added a test which is expected to fail (but does not) with a todo pointing to the issue. Will send a separate PR for fixing the issue.